Propose a SPIFFE Filesystem Delivery standard (#376)
spiffe/spiffe: standards/SPIFFE_Filesystem_Delivery.md

3\.2\. [Loading and Refreshing the Credential Bundle](#loading-and-refreshing-the-credential-bundle)
3\.3\. [Validating Peer SPIFFE Certificates](#validating-peer-spiffe-certificates)
Appendix A. [Filesystem Delivery and the Workload API](#appendix-filesystem-delivery-vs-the-workload-api)
Appendix B. [Example Kubernetes Setup](#appendix-example-kubernetes-setup)

## 1. Participants {#participants}

**Provisioning System:** The software stack responsible for distributing X.509
SVIDS and trust domain federation configuration to individual applications.

For applications running in a Kubernetes cluster, this might be composed of
kubelet, kube-apiserver, and a SPIFFE signer plugin, working through the Pod
Certificates mechanism to deliver certificates into individual application pod
filesystems.

For applications running directly on a machine or VM, this may be a system
daemon or node agent that writes X.509 SVIDs to a particular folder in the
filesystem.

**Application:** An individual application that expects to:
* Load an X.509 SVID from the filesystem, and use it to connect to other peers.
* Verify X.509 SVIDs presented by peers, including peers in different trust

>   domains.

Reviewer: We've decided on punting on this one by saying other files can exist in the directory without issue, and talking about standardizing more in the future. This would allow changes to be tested out in the mean time while the fd delivery spec is standardized as experimental. Should be revisited before GA. I cant seem to dismiss this issue, but I think it is resolved for now.

The application SHOULD search for a SPIFFE credential folder with the following
procedure:
1. The application should first follow the procedure from [SPIFFE Workload
   Endpoint, Section
   4](https://spiffe.io/docs/latest/spiffe-specs/spiffe_workload_endpoint/#4-locating-the-endpoint)
   to determine if the SPIFFE Workload API is available.
2. If the applications supports an application-specific configuration mechanism
   for picking a SPIFFE credential folder, it use the credential folder
   indicated by that configuration.
3. Otherwise, if the SPIFFE_CREDENTIAL_FOLDER environment variable is set, and
   contains a valid path for the platform, the application should treat that
   folder as the credential folder to use.
4. Otherwise, the application should check for the existence of a
   platform-specific default credential folder, and use it if it is present.
5. Otherwise, the application has not been issued SPIFFE credentials.

On platforms that use the Filesystem Hierarchy Standard, such as Linux, the
application should check the following default credential folders, in descending
order of preference:
1) `/run/secrets/workload-spiffe-credentials/credential-bundle.pem`

> 2) `/var/run/secrets/workload-spiffe-credentials/credential-bundle.pem`

Reviewer 2: Closing this as outdated.

  
On most systems, `/var/run` is a symlink to `/run`, or vice-versa.  However this
may not be true in some containerized environments, especially those that use
distroless base images.

### 3.2 Loading and Refreshing the Credential Bundle {#loading-and-refreshing-the-credential-bundle}

Before the application can use its SPIFFE credentials, it needs to load the

> credential bundle from the filesystem.

Reviewer: It can take a bit for the dir to populate. it being not fully baked at times I think is expected behavior. The app should retrry

When loading the credential bundle, the application SHOULD complete the load in
a single open/read/close cycle.  In particular, the application SHOULD NOT have
one open/read/close cycle to read the private key, and another open/read/close
cycle to read the certificate chain.

Note: If the application does use more than one open/read/close cycle, then it
is possible that the application may load a mismatched private key and
certificate, if the provisioning system rotated the credentials between the two
open/read/close cycles.  This is true even if the provisioning system atomically
writes the updated content.

The provisioning system may periodically update the credential bundle on the
filesystem.  The application SHOULD reload the credential bundle within five

> minutes of the provisioning system updating it.  The application SHOULD NOT

Reviewer 2: Discussed last sig-spec that we're removing the MUST and just keeping the "SHOULD reflect updates from the filesystem as soon as reasonably possible."

assume that the updated bundle will have any commonality with the previous
bundle.  For example, the type of the private key may be different, or the
certificate may be issued from a different root, with different intermediates.

> ### 3.3 Validating Peer SPIFFE Certificates {#validating-peer-spiffe-certificates}

Reviewer 3: Resolved - thank you.

3\.3\. [Validating Peer SPIFFE Certificates](#validating-peer-spiffe-certificates)
Appendix A. [Filesystem Delivery and the Workload API](#appendix-filesystem-delivery-vs-the-workload-api)
Appendix B. [Example Kubernetes Setup](#appendix-example-kubernetes-setup)

## 1. Participants {#participants}

**Provisioning System:** The software stack responsible for distributing X.509
SVIDS and trust domain federation configuration to individual applications.

For applications running in a Kubernetes cluster, this might be composed of
kubelet, kube-apiserver, and a SPIFFE signer plugin, working through the Pod
Certificates mechanism to deliver certificates into individual application pod
filesystems.

For applications running directly on a machine or VM, this may be a system
daemon or node agent that writes X.509 SVIDs to a particular folder in the
filesystem.

**Application:** An individual application that expects to:
* Load an X.509 SVID from the filesystem, and use it to connect to other peers.
* Verify X.509 SVIDs presented by peers, including peers in different trust
  domains.

## 2. Credential Folder {#credential-folder}

A credential folder contains:

> * A single credential bundle, named `credential-bundle.pem`

Reviewer: resolved

these updates atomically, so that an application reading a trust bundle reads
either the complete old content, or the complete new content, with no
intermediate state.

The certificates in a SPIFFE trust bundle file should only be used by
SPIFFE-aware applications.  It is not generally safe to attempt to use any of
these certificates for standard server TLS certificate verification.  In
particular, it is *never* safe to combine the contents of more than one SPIFFE
trust bundle into a single, undifferentiated bag of root certificates.

## 3. Application Behavior {#application-behavior}

### 3.1 Locating the Credential Folder {#locating-the-credential-folder}
The application SHOULD search for a SPIFFE credential folder with the following
procedure:
1. The application should first follow the procedure from [SPIFFE Workload
   Endpoint, Section
   4](https://spiffe.io/docs/latest/spiffe-specs/spiffe_workload_endpoint/#4-locating-the-endpoint)
   to determine if the SPIFFE Workload API is available.
2. If the applications supports an application-specific configuration mechanism
   for picking a SPIFFE credential folder, it use the credential folder
   indicated by that configuration.

> 3. Otherwise, if the SPIFFE_CREDENTIAL_FOLDER environment variable is set, and

Reviewer: We decided to move this out to its own spec. I think we just reference here the other spec, and that we're implementing file://<path to fs delivery directory>

   contains a valid path for the platform, the application should treat that
   folder as the credential folder to use.
4. Otherwise, the application should check for the existence of a

>    platform-specific default credential folder, and use it if it is present.

Reviewer: Cant seem to dismiss this. It seems to be resolved.

5. Otherwise, the application has not been issued SPIFFE credentials.

On platforms that use the Filesystem Hierarchy Standard, such as Linux, the
application should check the following default credential folders, in descending
order of preference:

> 1) `/run/secrets/workload-spiffe-credentials/credential-bundle.pem`

Reviewer: Resolved.

2) `/var/run/secrets/workload-spiffe-credentials/credential-bundle.pem`
  
On most systems, `/var/run` is a symlink to `/run`, or vice-versa.  However this
may not be true in some containerized environments, especially those that use
distroless base images.

### 3.2 Loading and Refreshing the Credential Bundle {#loading-and-refreshing-the-credential-bundle}

Before the application can use its SPIFFE credentials, it needs to load the
credential bundle from the filesystem.

When loading the credential bundle, the application SHOULD complete the load in
a single open/read/close cycle.  In particular, the application SHOULD NOT have
one open/read/close cycle to read the private key, and another open/read/close
cycle to read the certificate chain.

Note: If the application does use more than one open/read/close cycle, then it
is possible that the application may load a mismatched private key and
certificate, if the provisioning system rotated the credentials between the two
open/read/close cycles.  This is true even if the provisioning system atomically
writes the updated content.

> The provisioning system may periodically update the credential bundle on the

Reviewer: I'd still add in, that the applications should only read the files via:
open, read entire file, close, without seeking.

This allows implementations that would normally not be allowed if that wasnt a requirement.

> The provisioning system may periodically update the credential bundle on the

Reviewer: Say your backing the credential directory with spiffefs and it crashes and gets restarted. there is a time where the mount may read empty, but then gets re-populated. Does the app drop in this case, or just wait a bit and recover?

> The provisioning system may periodically update the credential bundle on the

Reviewer: What if we make the signal more explicit? truncate the file to 0 in case we want to signal it is no longer to be used?

File gone means keep using the content you have in memory until it expires, you exit or you see a  new one show back up. File truncated, explicit, remove the svid as soon as you detect it

filesystem.  The application SHOULD reload the credential bundle as soon as
reasonably possible after the provisioning system updating it.  The application
SHOULD NOT assume that the updated bundle will have any commonality with the
previous bundle.  For example, the type of the private key may be different, or
the certificate may be issued from a different root, with different

> intermediates.

Reviewer: Resolved.

### 3.3 Validating Peer SPIFFE Certificates {#validating-peer-spiffe-certificates}

In order to verify a SPIFFE X.509 SVID presented by a peer, the application
needs to determine the correct SPIFFE trust bundle for the peer's trust domain.

Once the application has parsed the peer's trust domain from the unverified peer
X.509 SVID, the application should load the corresponding trust domain's
trust bundle from the credential folder.  If no trust bundle file is present for
the peer's trust domain, the application SHOULD assume that SPIFFE federation
has not been configured with the peer, and fail to verify the peer's
certificate.

The provisioning system may periodically update the number and contents of the
per-trust-domain trust bundles in the filesystem.  The application MAY cache the
presence, absence, and contents of each trust domain's trust bundle in memory,
but it SHOULD reflect updates from the filesystem as soon as reasonably
possible.

The application SHOULD handle de-federation; if it has loaded a trust bundle for
trust domain X, and then X's trust bundle is removed from the credential folder,
the application SHOULD begin failing to verify certificates for trust domain X

> as soon as reasonably possible.  This only needs to affect new TLS handshakes;

Reviewer: I agree it should be standardized behavior. Not sure it belongs in this spec though. This is delivering federated bundle info to the app. the behavior of the app should be consistent between fd delivery and workload-api delivery. So maybe in the federation standards docs?

> as soon as reasonably possible.  This only needs to affect new TLS handshakes;

Reviewer 2: We discussed last sig-spec to remove that specific sentence about existing TLS sessions and instead tackle it SPIFFE overall as it also applies to SPIFFE Workload API delivery.

> as soon as reasonably possible.  This only needs to affect new TLS handshakes;

Reviewer 2: Created https://github.com/spiffe/spiffe/issues/424

predates Filesystem Delivery, and is the preferred solution, all else equal.

Filesystem Delivery exists because, in certain computing environments, such as
Kubernetes, it is easier for the environment to furnish *files* to the workload,
rather than making a TCP or Unix socket service available.  It's also lower
effort to read files from workloads, but, as described in [Section
3.2](#loading-and-refreshing-the-credential-bundle), it still requires care.

Given these two choices, which should you use?

If you are writing or maintaining an application that needs to load and use
SPIFFE credentials, you should ideally support both the Workload API and
Filesystem Delivery, following the procedure from [Section
3.1](#locating-the-credential-folder) to select.

If supporting both is not feasible, then consider which standard the SPIFFE
identity provisioning you are most likely to use supports.

If you are writing or maintaining a SPIFFE X.509 identity provisioning system,
then, again, ideally you would support both mechanisms, however (as in the case
of Kubernetes) the Workload API may be less feasible.

## Appendix B: Example Kubernetes Setup {#example-kubernetes-setup}

> 

Reviewer 2: Discussed last sig-spec that the appendix is explicitly non-normative and with that an example only. We may come back to this and have a specific best practices or other document that describe the usage in Kubernetes in more detail.

  automatically updates as the Cluster Trust Bundles are updated.

For example, we could implement a controller that handles the signer name
`spiffe.example/identity`.  This controller would have two responsibilities:
* Respond to PodCertificateRequests from pods that want a certificate from the
  `spiffe.example/identity`.  The precise details of the issued certificate are
  up to the controller.  In this case, assume that it is configured to issue
  certificates with SPIFFE IDs like
  `spiffe://domain-a.myorg.example/ns/<namespace>/sa/<service-account>`.
* Maintain a set of ClusterTrustBundles associated with
  `spiffe.example/identity`.  There are many potential ways to structure the set
  of ClusterTrustBundles, but one that would work is to use labels to divide
  them:
    * `k8s.spiffe.io/canarying`: Possible values of `live` or `preview`.  Most
      application pods would always select ClusterTrustBundles with the value
      `live`, but some may be configured to select `preview`, so that actions
      like rotating a CA or federating with a new trust domain can be previewed
      on some fraction of your application fleet.

>     * `k8s.spiffe.io/workload-trust-domain`:  The label's value is a trust

Reviewer 2: Discussed last sig-spec that the appendix is explicitly non-normative and with that an example only. We may come back to this and have a specific best practices or other document that describe the usage in Kubernetes in more detail.

      domain X; workloads that are part of X should load this trust bundle.

>     * `k8s.spiffe.io/peer-trust-domain`: The label's value is a trust domain Y;

Reviewer 2: Discussed last sig-spec that the appendix is explicitly non-normative and with that an example only. We may come back to this and have a specific best practices or other document that describe the usage in Kubernetes in more detail.

> # The SPIFFE Filesystem Delivery Standard

Reviewer 3: Prior to merging, we should add the appropriate STABILITY header as per https://github.com/spiffe/spiffe/blob/main/standards/STABILITY.md

@arndt-s / @sorindumitru - what level do we intend to land this at? I think I'd be happy with Incubating (or Experimental) but ideally not Stable at this time.

Reviewer: I'd vote Experimental. There are many details we decided to punt on to try and expedite this. It is not complete enough for Incubating IMO. (Ex: multiple certs, hints, types other then x509)

Reviewer 2: I think `Incubating` is appropriate here.
> The criterion for promotion to Incubating is SIG-Spec consensus that the design is sound and ready for real world scenarios.

Reviewer: Hmm.. rereading through the stability spec, I'd agree with Incubating once the x509 filename prefix makes it in. The other details to work out should be implementable entirely with additions rather then changes.

Reviewer 2: Added STABILITY header with `Incubating`

3\.3\. [Validating Peer SPIFFE Certificates](#validating-peer-spiffe-certificates)
Appendix A. [Filesystem Delivery and the Workload API](#appendix-filesystem-delivery-vs-the-workload-api)
Appendix B. [Example Kubernetes Setup](#appendix-example-kubernetes-setup)

## 1. Participants {#participants}

**Provisioning System:** The software stack responsible for distributing X.509
SVIDS and trust domain federation configuration to individual applications.

For applications running in a Kubernetes cluster, this might be composed of
kubelet, kube-apiserver, and a SPIFFE signer plugin, working through the Pod
Certificates mechanism to deliver certificates into individual application pod
filesystems.

For applications running directly on a machine or VM, this may be a system
daemon or node agent that writes X.509 SVIDs to a particular folder in the
filesystem.

**Application:** An individual application that expects to:
* Load an X.509 SVID from the filesystem, and use it to connect to other peers.
* Verify X.509 SVIDs presented by peers, including peers in different trust
  domains.

## 2. Credential Folder {#credential-folder}

A credential folder contains:

> * A single credential bundle, named `credential-bundle.private-key.pem`

Reviewer: ```suggestion
* A single credential bundle, named `x509.credential-bundle.private-key.pem`
```

Reviewer 2: I disagree with this logic. I strongly suggest keeping the scope to a single X.509 SVID for now.

Reviewer: We chatted in sig spec today. Dropping the multiple ID thing but keeping the x509 bit in the proposal and then letting multi ID cook outside of the spec. amended

> * One or more trust bundles, named `<trust-domain>.spiffe-trust-bundle.pem`

Reviewer: ```suggestion
* One or more trust bundles, named `x509.<trust-domain>.spiffe-trust-bundle.pem`
```

Reviewer 4: I still don't really like the prefix, but if we do add it, I think `<trust-domain>.x509.spiffe-trust-bundle.pem` makes it slightly nicer to parse out the trust domain.

Reviewer: I'm good with that format.

Reviewer: how do you propose we handle WIMSE without it?

Reviewer 2: Resolved in last sig-spec, we settled on using a `.x509.pem` postfix

knows where to find the additional credential folders, and in which
circumstances they should be used.

Because the credential folder contains sensitive credentials, provisioning
systems SHOULD ensure that their contents do not appear on physical disks or in
automatic backups.

### 2.1 Credential Bundle {#credential-bundle}

A credential bundle file contains the application's private key and certificate
chain.  Combining these together into one file ensures that the provisioning
system can rotate the key and certificate atomically.  (Note that, for safety,
the application must also read this file atomically, as described in section
3.2).

The credential bundle consists of two or more PEM blocks.  The first block must
be of type PRIVATE KEY, and contain a PKCS#8-serialized private key.

Compatibility Note: Some provisioning systems offer features for keeping the
private key inaccessible to the application, only exposing RPCs for signing
operations over the private key.  When a feature like this is in use, the
credential bundle might omit the private key.  The provisioning system and
application must be coordinated to make this happen.  SPIFFE-compliant

> applications are not required to seamlessly handle this case.

Reviewer: This paragraph parses out to "some apps may not follow this standard" ?

The remaining PEM blocks must be of type CERTIFICATE, and contain the
application's certificate chain, in leaf-to-root order.  The leaf certificate
must be issued to the public key derived from the PRIVATE KEY block.  The leaf
certificate SHOULD be an X.509 SVID.

The certificate chain MUST NOT contain the trust anchor ("root certificate") of
the chain.  It MUST stop at the certificate preceding the trust anchor.

The credential bundle:
* MUST NOT contain any other PEM block types.
* MUST NOT contain any inter-block data.
* MUST NOT contain any PEM blocks that have PEM headers.

> 

Reviewer: ```suggestion
* MUST be read only linearity, then the file must be closed and reopened before reading again.
 
```

The provisioning system SHOULD update the credential bundle on the filesystem
with a new key and certificate chain before the leaf certificate currently in
the bundle expires, with enough additional time for the application to detect
and reload the updated credential bundle. The provisioning system SHOULD make
these updates atomically, so that an application reading the credential bundle
reads either the complete old content or the complete new content, with no
intermediate state.

The developer SHOULD NOT copy this file anywhere, since compromise of the
credential-bundle.private-key.pem file could compromise the security of TLS
sessions. The provisioning system SHOULD be configured to limit access to the
credential-bundle.private-key.pem file to only the necessary user(s).

### 2.2 SPIFFE Trust Bundles {#spiffe-trust-bundles}

The credential folder will contain one or more trust bundles, one per trust
domain that the provisioning system is configured to federate with.

Each trust bundle file MUST be named following the pattern

> "<trust-domain>.spiffe-trust-bundle.pem".  The file contents are one or more PEM

Reviewer: ```suggestion
"x509.<trust-domain>.spiffe-trust-bundle.pem".  The file contents are one or more PEM
```

system's federation configuration changes.  The provisioning system SHOULD make
these updates atomically, so that an application reading a trust bundle reads
either the complete old content, or the complete new content, with no
intermediate state.

The certificates in a SPIFFE trust bundle file should only be used by
SPIFFE-aware applications.  It is not generally safe to attempt to use any of
these certificates for standard server TLS certificate verification.  In
particular, it is *never* safe to combine the contents of more than one SPIFFE
trust bundle into a single, undifferentiated bag of root certificates.

## 3. Application Behavior {#application-behavior}

### 3.1 Locating the Credential Folder {#locating-the-credential-folder}
The application SHOULD search for a SPIFFE credential folder with the following
procedure:
1. The application should first follow the procedure from [SPIFFE Workload
   Endpoint, Section
   4](https://spiffe.io/docs/latest/spiffe-specs/spiffe_workload_endpoint/#4-locating-the-endpoint)
   to determine if the SPIFFE Workload API is available.
2. If the application supports an application-specific configuration mechanism

>    for picking a SPIFFE credential folder, it should use the credential folder

Reviewer 4: ```suggestion
   for picking a SPIFFE credential folder, it SHOULD use the credential folder
```

   indicated by that configuration.

> 3. Otherwise, if the SPIFFE_CREDENTIAL_FOLDER environment variable is set, and

Reviewer 4: It might be best to remove this point for now while we discuss the implementation of SPIFFE_ENDPOINT.

   contains a valid path for the platform, the application should treat that
   folder as the credential folder to use.
4. Otherwise, the application has not been issued SPIFFE credentials.

This specification does not define a default or well-known location for the
credential folder.

### 3.2 Loading and Refreshing the Credential Bundle {#loading-and-refreshing-the-credential-bundle}

Before the application can use its SPIFFE credentials, it needs to load the
credential bundle from the filesystem.

> When loading the credential bundle, the application SHOULD complete the load in

Reviewer 4: SHOULD can lead to some application errors, so let's do MUST.
```suggestion
When loading the credential bundle, the application MUST complete the load in
```

> a single open/read/close cycle.  In particular, the application SHOULD NOT have

Reviewer: ```suggestion
a single open/read/close cycle.  In particular, the application MUST NOT have
```

one open/read/close cycle to read the private key, and another open/read/close
cycle to read the certificate chain.

> Note: If the application does use more than one open/read/close cycle, then it
> is possible that the application may load a mismatched private key and
> certificate, if the provisioning system rotated the credentials between the two
> open/read/close cycles.  This is true even if the provisioning system atomically
> writes the updated content.

Reviewer: ```suggestion
```

Lets kill this and just make it required. When implementing spiffefs, it requires linear reading. It does not support seeking. So clients can only ever read it linearly. Its easy enough to have any client read a while pem file into emory and process it.

The provisioning system may periodically update the credential bundle on the

> filesystem.  The application SHOULD reload the credential bundle as soon as

Reviewer 4: I think this also should be a MUST, not doing it leads to errors.
```suggestion
filesystem.  The application MUST reload the credential bundle as soon as
```

> reasonably possible after the provisioning system updating it.  The application
> MUST reload the credential bundle within 5 minutes of an update on the
> filesystem.  The application SHOULD NOT assume that the updated bundle will have

Reviewer 4: The 5 minute interval also depends on the TTL of the cert. Likely only important for testing, but still worth pointing out:

```suggestion
reasonably possible after the provisioning system updating it.  The application MUST reload the credential bundle as soon as easonably possible after the provisioning system updating it.  If not dictated otherwise by application configuration, the application MUST refresh the credential bundle at least every 5 minutes or half of the TTL of the leaf certificate, whichever is shorter. The application SHOULD NOT assume that the updated bundle will have
```

Reviewer: I think the idea here is that if the implementation doesnt have an inotify like mechanism, it would be semi reasonable for it to just poll at ~5 min intervals?

Reviewer 4: min(5 minutes, TTL/2), I'd say.

Reviewer: hmmm... there is nothing providing ttl's to the app.

It could simulate it by looking at the cert and doing ttl=endtime - starttime?

Reviewer 2: Discussed last sig-spec to remove this specific sentence.

any commonality with the previous bundle.  For example, the type of the private
key may be different, or the certificate may be issued from a different root,
with different intermediates.

### 3.3 Validating Peer SPIFFE Certificates {#validating-peer-spiffe-certificates}

In order to verify a SPIFFE X.509 SVID presented by a peer, the application
needs to determine the correct SPIFFE trust bundle for the peer's trust domain.

Once the application has parsed the peer's trust domain from the unverified peer
X.509 SVID, the application should load the corresponding trust domain's
trust bundle from the credential folder.  If no trust bundle file is present for
the peer's trust domain, the application SHOULD assume that SPIFFE federation
has not been configured with the peer, and fail to verify the peer's
certificate.

The provisioning system may periodically update the number and contents of the
per-trust-domain trust bundles in the filesystem.  The application MAY cache the
presence, absence, and contents of each trust domain's trust bundle in memory,
but it SHOULD reflect updates from the filesystem as soon as reasonably

> possible.

Reviewer 4: ```suggestion
possible. It is recommended to reload the bundle at least every 5 minutes since the bundle can change at any time.
```

The application SHOULD handle de-federation; if it has loaded a trust bundle for
trust domain X, and then X's trust bundle is removed from the credential folder,
the application SHOULD begin failing to verify certificates for trust domain X
as soon as reasonably possible.  This only needs to affect new TLS handshakes;
established TLS connections may remain open until they are otherwise finished or
interrupted.

As mentioned in section 2.2, it is not safe to use the contents of a SPIFFE
trust bundle file except by a SPIFFE-aware application performing SPIFFE X.509
SVID verification for a certificate from the corresponding trust domain.
Applications should NEVER unify the contents of multiple

> `*.spiffe-trust-bundle.pem` files into a single root store for certificate

Reviewer: ```suggestion
`x509.*.spiffe-trust-bundle.pem` files into a single root store for certificate
```

because they happen to run in the same Kubernetes cluster.

This information can then be mounted into an application pod in a way that forms
a valid SPIFFE credential folder, with the SPIFFE_CREDENTIAL_FOLDER environment
variable pointing the application at the mount path:

```
apiVersion: apps/v1
kind: Deployment
metadata:
  name: client
  namespace: spiffe-example
  labels:
    app: client
spec:
  replicas: 1
  selector:
    matchLabels:
      app: client
  template:
    metadata:
      labels:
        app: client
    spec:
      restartPolicy: Always
      containers:
      - name: client
        image: debian
        command: ["sleep", "infinity"]
        env:
        - name: SPIFFE_CREDENTIAL_FOLDER
          value: /run/workload-spiffe-credentials
        volumeMounts:
        - name: spiffe-credentials
          mountPath: /run/workload-spiffe-credentials
          readOnly: true
      volumes:
      - name: spiffe-credentials
        projected:
          sources:
          - clusterTrustBundle:
              signerName: spiffe.example/identity
              labelSelector:
                matchLabels:

>                   "k8s.spiffe.io/canarying":             "live"
>                   "k8s.spiffe.io/workload-trust-domain": "a.myorg.example"
>                   "k8s.spiffe.io/peer-trust-domain":     "a.myorg.example"
>               path: a.myorg.example.spiffe-trust-bundle.pem

Reviewer 4: nit: maybe `spiffe.io` is enough for this. We also do not support canarying (and I'm not sure what the signer or workload would do with that information) so that seems unnecessary.

As discussed on Slack, I'm not sure we need both workload-trust-domain and peer-trust-domain. The workload just has a folder with trust bundles for various trust domains and it trusts all of them. Maybe it would be easier to specify that a trust domain is federated with a bool? Let me know what you think.

```suggestion
               "spiffe.io/trust-domain":     "a.myorg.example"
              path: a.myorg.example.spiffe-trust-bundle.pem
```

Reviewer 2: Discussed in sig-spec that we keep appendix as is (non-normative).

>               path: a.myorg.example.spiffe-trust-bundle.pem

Reviewer: ```suggestion
              path: x509.a.myorg.example.spiffe-trust-bundle.pem
```

          # This workload needs to federate with b.myorg.example, so it needs the appropriate bundle.
          - clusterTrustBundle:
              signerName: spiffe.example/identity
              labelSelector:
                matchLabels:
                  "k8s.spiffe.io/canarying":             "live"
                  "k8s.spiffe.io/workload-trust-domain": "a.myorg.example"
                  "k8s.spiffe.io/peer-trust-domain":     "b.myorg.example"

>               path: b.myorg.example.spiffe-trust-bundle.pem

Reviewer: ```suggestion
              path: x509.b.myorg.example.spiffe-trust-bundle.pem
```

          - podCertificate:
              signerName: spiffe.example/identity
              keyType: ECDSAP256

>               credentialBundlePath: credential-bundle.pem

Reviewer: ```suggestion
              credentialBundlePath: x509.credential-bundle.private-key.pem
```

Reviewer: Suggestion edited as per sig spec discussion
